source/common/network/listen_socket_impl.h

```diff
@@ -50,5 +50,63 @@ class UdsListenSocket : public ListenSocketImpl {
   UdsListenSocket(const std::string& uds_path);
 };
 
+class ConnectionSocketImpl : virtual public ConnectionSocket {
+public:
+  ConnectionSocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address,
+                       const Address::InstanceConstSharedPtr& remote_address)
+      : fd_(fd), local_address_(local_address), remote_address_(remote_address) {}
+  ~ConnectionSocketImpl() { close(); }
+
+  // Network::ConnectionSocket
+  const Address::InstanceConstSharedPtr& localAddress() const override { return local_address_; }
+  const Address::InstanceConstSharedPtr& remoteAddress() const override { return remote_address_; }
+  bool localAddressReset() const override { return local_address_reset_; }
+  int fd() const override { return fd_; }
+  void close() override {
+    if (fd_ != -1) {
+      ::close(fd_);
+      fd_ = -1;
+    }
+  }
+
+protected:
+  int fd_;
+  Address::InstanceConstSharedPtr local_address_;
+  Address::InstanceConstSharedPtr remote_address_;
+  bool local_address_reset_{false};
+};
+
+class AcceptedSocketImpl : public AcceptedSocket, public ConnectionSocketImpl {
+public:
+  AcceptedSocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address,
+                     const Address::InstanceConstSharedPtr& remote_address)
+      : ConnectionSocketImpl(fd, local_address, remote_address) {}
+  ~AcceptedSocketImpl() {}
+
+  // Network::AcceptedSocket
+  void resetLocalAddress(const Address::InstanceConstSharedPtr& local_address) override {
+    if (*local_address != *local_address_) {
```

Comment:
nit: The comparison here is almost definitely slower than just assigning universally. Or is this to prevent reset from being set? If so comment on how/why/when this happens?

Reply:
It was to prevent the flag being set, but I refactored this and made it the callers responsibility to make sure the addresses are actually different when that matters.